Make verification per-ciphersuite

conradoplg opened the pull request:
Based on #96 

Closes #91 

As it stands, the RFC can be interpreted as basically requiring cofactored verification (the cofactor curves require it, and the curves without cofactor don't, so we can always just multiply by the factor even if it's 1 and it will cover both cases). Thus the current implementation already follows the RFC. But in case there is the need to override it, this PR:

- Adds `VerifySignature` method to the `Ciphersuite` trait, with a default implementation with cofactored verification
- Changes the existing `VerifyingKey.verify()` to call that

There are many possible designs here, this is just what seemed to make more sense. We could also:

- Do nothing
- Don't provide a default implementation, which would allow us to remove `Group::cofactor()` if we want, but then we'd need to add one to every ciphersuite

dconnolly commented on frost-core/src/lib.rs:
```diff
@@ -191,6 +191,32 @@ pub trait Ciphersuite: Copy + Clone {
+    /// equation (it multiplies by the cofactor returned by [`Group::cofactor()`]).
+    /// You may override this to provide a tailored implementation, but it must also multiply
+    /// by the cofactor to comply with the RFC.
+    fn VerifySignature(
```
We can match the standard Rust naming for this function, like `generate()`, etc

```suggestion
    fn verify_signature(
```

conradoplg replied:
done in 27a5b93

dconnolly commented on frost-core/src/verifying_key.rs:
```diff
@@ -35,21 +35,7 @@ where
-        } else {
-            Err(Error::InvalidSignature)
-        }
+        C::VerifySignature(msg, signature, self)
```
👍

dconnolly commented on frost-core/src/lib.rs:
```diff
@@ -191,6 +191,32 @@ pub trait Ciphersuite: Copy + Clone {
+
+    /// Verify a signature for this ciphersuite. The default implementation uses the "cofactored"
+    /// equation (it multiplies by the cofactor returned by [`Group::cofactor()`]).
+    /// You may override this to provide a tailored implementation, but it must also multiply
+    /// by the cofactor to comply with the RFC.
```
```suggestion
    /// You may override this to provide a tailored implementation, but if the ciphersuite defines it, it must also multiply
    /// by the cofactor to comply with the RFC.
```

conradoplg replied:
done in 27a5b93

dconnolly reviewed (commented):
A few nits, looks great 👍

dconnolly commented on frost-core/src/lib.rs:
```diff
@@ -191,6 +191,32 @@ pub trait Ciphersuite: Copy + Clone {
+
+    /// Verify a signature for this ciphersuite. The default implementation uses the "cofactored"
+    /// equation (it multiplies by the cofactor returned by [`Group::cofactor()`]).
+    /// You may override this to provide a tailored implementation, but if the ciphersuite defines it,
+    /// it must also multiply by the cofactor to comply with the RFC.
+    fn verify_signature(
```
```suggestion
    /// # Cryptographic Safety
    /// 
    /// You may override this to provide a tailored implementation, but if the ciphersuite defines it,
    /// it also multiply by the cofactor to comply with the RFC.
    fn verify_signature(
```

conradoplg commented:
I fixed the conflicts, but I'm not sure how to handle `verify_prehashed`. I moved it to a crate function that is now called by the default implementation of `verify_signature`. But this means that if the ciphersuite provides a custom implementation of `verify_signature`, then the batch verifier won't use it.

The alternative would be to add `verify_prehashed` to the ciphersuite. @dconnolly what do you think?

dconnolly commented:
Aaah good point. Hm. 

Could duplicate this somewhat. 

For now, for first-pass support, we can just have a TODO that basically says "batch verification supports Schnorr-ish verification equations and does the cofactor multiplication, if your concrete impl of Ciphersuite has a verify_signature() that diverges, you may not want to also use the batch math

dconnolly commented:
> Aaah good point. Hm. 
> 
> Could duplicate this somewhat. 
> 
> For now, for first-pass support, we can just have a TODO that basically says "batch verification supports Schnorr-ish verification equations and does the cofactor multiplication, if your concrete impl of Ciphersuite has a verify_signature() that diverges, you may not want to also use the batch math

So that may mean we put batch support behind feature flag

conradoplg commented:
> Aaah good point. Hm.
> 
> Could duplicate this somewhat.
> 
> For now, for first-pass support, we can just have a TODO that basically says "batch verification supports Schnorr-ish verification equations and does the cofactor multiplication, if your concrete impl of Ciphersuite has a verify_signature() that diverges, you may not want to also use the batch math

Sounds good, added a comment in a645a5b
